GameRoom.ts

```diff
@@ -1,500 +1,637 @@
-                            } else {
-                                console.error('failed to generate random Layout');
-                            }
-                        }
-                      try {
-                    } catch (err) {
-                        console.error('an error occurred while trying to generate the Field', err);
-                    }
-                }),
-                (reason: any) => {
-                  console.error("failed, couldn't get TileSet.fields", reason);
-                })
-            }),
-            (reason: any) => {
-              console.error("failed, couldn't get TileSet", reason);
-            });
-
-
-        this.state.physicsState.setBroadcastCallback(this.broadcast.bind(this));
-        this.state.physicsState.addDice();
-        this.state.physicsState.setOnDiceThrow(((num: number) => {
-            if (this.state.action === Actions[Actions.ROLL]) {
-                this.state.nextAction(this);
-            }
-        }).bind(this));
-
-        this.onMessage(MessageType.CHAT_MESSAGE, this.onChatMessage.bind(this));
-        this.onMessage(MessageType.JOIN_MESSAGE, this.onJoinMessage.bind(this));
-        this.onMessage(MessageType.GAME_MESSAGE, this.onGameMessage.bind(this));
-        this.onMessage(MessageType.PLAYER_MESSAGE, this.onPlayerMessage.bind(this));
-        this.onMessage(MessageType.PHYSICS_MESSAGE, this.onPhysicsMessage.bind(this));
-        this.onMessage(MessageType.DEBUG_COMMAND, this.onDebugMessage.bind(this));
-        this.onMessage(MessageType.LEFT_MESSAGE, this.onLeftMessage.bind(this));
-        this.onMessage(MessageType.OTHER, this.onOtherMessage.bind(this));
-        this.onMessage(MessageType.REFRESH_COMMAND, this.onRefreshMessage.bind(this));
-        this.onMessage(MessageType.ACHIEVEMENT_MESSAGE, this.onAchievementMessage.bind(this));
-        this.onMessage(MessageType.ITEM_MESSAGE, this.onItemMessage.bind(this));
-        this.onMessage(MessageType.CHAT_COMMAND, this.onChatCommand.bind(this));
-
-        return undefined;
-    }
-
-    async onDispose(): Promise<void | Promise<any>> {
-        console.log(`[onDispose] Destructing physicsState`);
-        
-        const metadata = this.metadata as {roomName, author, skin, randomizeTiles};
-
-        await GameController.saveGameLog(
-            new Game(metadata.roomName, metadata.author, metadata.skin, metadata.randomizeTiles, this.createTime, new Date(), this.maxClients, this.state.round),
-            this.state.playerList,
-            this.createTime,
-            this.state.round)
-
-        this.state.physicsState.destructState();
-        return undefined;
+export class GameRoom extends Room<GameState, Metadata> {
+  createTime: Date;
+
+  /**
+   * will be called when a new room should be created
+   * @param options should be of type {@link CreateRoomOpts}
+   * @throws Error when type of options is incorrect
+   */
+  onCreate(options: any): void | Promise<any> {
```

Comment:
The nesting of promises seems a little messy is it possible to make this function asyc? or is this prohibited by the callback sgnature?